# Benchmark
[Vector Stores](https://python.langchain.com/docs/integrations/vectorstores)

In [6]:
import markdown_splitter as mdS
from langchain_community.embeddings import GPT4AllEmbeddings

docs = mdS.create_db(path='Folder', glob="**/*.md",split=None)

embeddings = GPT4AllEmbeddings()

query = "How install the chatbots"

repeat = 100

100%|██████████| 26/26 [00:00<00:00, 1055.51it/s]


bert_load_from_file: gguf version     = 2
bert_load_from_file: gguf alignment   = 32
bert_load_from_file: gguf data offset = 695552
bert_load_from_file: model name           = BERT
bert_load_from_file: model architecture   = bert
bert_load_from_file: model file type      = 1
bert_load_from_file: bert tokenizer vocab = 30522


## DeepLake

In [24]:
%%time 

# Store
from langchain_community.vectorstores import DeepLake
docsearch = DeepLake.from_documents(docs, embeddings)

Deep Lake Dataset in ./deeplake/ already exists, loading from the storage


Creating 69 embeddings in 1 batches of size 69:: 100%|██████████| 1/1 [00:04<00:00,  4.48s/it]

Dataset(path='./deeplake/', tensors=['embedding', 'id', 'metadata', 'text'])

  tensor      htype      shape      dtype  compression
  -------    -------    -------    -------  ------- 
 embedding  embedding  (138, 384)  float32   None   
    id        text      (138, 1)     str     None   
 metadata     json      (138, 1)     str     None   
   text       text      (138, 1)     str     None   
CPU times: user 19.5 s, sys: 719 ms, total: 20.2 s
Wall time: 5.71 s


In [25]:
%%time
for i in range(repeat):
    docsearch.similarity_search(query)

docsearch.similarity_search(query)

CPU times: user 21.7 s, sys: 4.06 s, total: 25.8 s
Wall time: 5.98 s


[Document(page_content="The chatbots are in their respective folders with an application launcher `autorun.sh` to install each of them without specific knowledge.  \nOn Mac, open the terminal and type:\n```shell\ncd\n```\nDrag the **`folder`** containing the file `autorun.sh`, then press the Enter key (↩︎).  \n_If you have done it correctly, the **`~`** between your machine's name (`name@MacBook-Pro-of-Name`) and the **`%`** sign should display the name of the `folder` instead._  \nExecute the following line of code by pressing the Enter key (↩︎):\n```shell\nsh autorun.sh\n```\nWait a moment, the model should open in your default web browser.", metadata={'Header 1': 'Onboarding Bot Model', 'Header 2': 'Installation'}),
 Document(page_content="The chatbots are in their respective folders with an application launcher `autorun.sh` to install each of them without specific knowledge.  \nOn Mac, open the terminal and type:\n```shell\ncd\n```\nDrag the **`folder`** containing the file `autoru

## Annoy

In [26]:
%%time 

# Store
from langchain_community.vectorstores import Annoy
docsearch = Annoy.from_documents(docs, embeddings)

CPU times: user 25.8 s, sys: 652 ms, total: 26.5 s
Wall time: 6.7 s


In [27]:
%%time
for i in range(repeat):
    docsearch.similarity_search(query)

docsearch.similarity_search(query)

CPU times: user 9.45 s, sys: 104 ms, total: 9.55 s
Wall time: 2.57 s


[Document(page_content="The chatbots are in their respective folders with an application launcher `autorun.sh` to install each of them without specific knowledge.  \nOn Mac, open the terminal and type:\n```shell\ncd\n```\nDrag the **`folder`** containing the file `autorun.sh`, then press the Enter key (↩︎).  \n_If you have done it correctly, the **`~`** between your machine's name (`name@MacBook-Pro-of-Name`) and the **`%`** sign should display the name of the `folder` instead._  \nExecute the following line of code by pressing the Enter key (↩︎):\n```shell\nsh autorun.sh\n```\nWait a moment, the model should open in your default web browser.", metadata={'Header 1': 'Onboarding Bot Model', 'Header 2': 'Installation'}),
 Document(page_content='The [Dev Launchers](https://devlaunchers.org) website is a community of aspiring developers eager to learn and gain experience. As such, projects are open-source and ambitious, allowing members to enhance their skills.  \nThe [ChatBot](https://en.

## AwaDB
! Look transformers, sentence_transformers and huggingface old versions

## Chroma

In [28]:
%%time 

# Store
from langchain_community.vectorstores import Chroma
docsearch = Chroma.from_documents(docs, embeddings)

CPU times: user 21.9 s, sys: 463 ms, total: 22.4 s
Wall time: 5.22 s


In [29]:
%%time
for i in range(repeat):
    docsearch.similarity_search(query)

docsearch.similarity_search(query)

CPU times: user 11 s, sys: 150 ms, total: 11.1 s
Wall time: 3.15 s


[Document(page_content="The chatbots are in their respective folders with an application launcher `autorun.sh` to install each of them without specific knowledge.  \nOn Mac, open the terminal and type:\n```shell\ncd\n```\nDrag the **`folder`** containing the file `autorun.sh`, then press the Enter key (↩︎).  \n_If you have done it correctly, the **`~`** between your machine's name (`name@MacBook-Pro-of-Name`) and the **`%`** sign should display the name of the `folder` instead._  \nExecute the following line of code by pressing the Enter key (↩︎):\n```shell\nsh autorun.sh\n```\nWait a moment, the model should open in your default web browser.", metadata={'Header 1': 'Onboarding Bot Model', 'Header 2': 'Installation'}),
 Document(page_content="The chatbots are in their respective folders with an application launcher `autorun.sh` to install each of them without specific knowledge.  \nOn Mac, open the terminal and type:\n```shell\ncd\n```\nDrag the **`folder`** containing the file `autoru

## FAISS

In [30]:
%%time 

# Store
from langchain_community.vectorstores import FAISS
docsearch = FAISS.from_documents(docs, embeddings)

CPU times: user 23.3 s, sys: 536 ms, total: 23.8 s
Wall time: 5.73 s


In [31]:
%%time
for i in range(repeat):
    docsearch.similarity_search(query)

docsearch.similarity_search(query)

CPU times: user 9.49 s, sys: 107 ms, total: 9.59 s
Wall time: 2.47 s


[Document(page_content="The chatbots are in their respective folders with an application launcher `autorun.sh` to install each of them without specific knowledge.  \nOn Mac, open the terminal and type:\n```shell\ncd\n```\nDrag the **`folder`** containing the file `autorun.sh`, then press the Enter key (↩︎).  \n_If you have done it correctly, the **`~`** between your machine's name (`name@MacBook-Pro-of-Name`) and the **`%`** sign should display the name of the `folder` instead._  \nExecute the following line of code by pressing the Enter key (↩︎):\n```shell\nsh autorun.sh\n```\nWait a moment, the model should open in your default web browser.", metadata={'Header 1': 'Onboarding Bot Model', 'Header 2': 'Installation'}),
 Document(page_content='The [Dev Launchers](https://devlaunchers.org) website is a community of aspiring developers eager to learn and gain experience. As such, projects are open-source and ambitious, allowing members to enhance their skills.  \nThe [ChatBot](https://en.

## LanceDB

In [32]:
%%time 

# Store
from langchain.vectorstores import LanceDB
docsearch = LanceDB.from_documents(docs, embeddings)

CPU times: user 20.4 s, sys: 413 ms, total: 20.8 s
Wall time: 4.69 s


In [33]:
%%time
for i in range(repeat):
    docsearch.similarity_search(query)

docsearch.similarity_search(query)

CPU times: user 12.3 s, sys: 299 ms, total: 12.6 s
Wall time: 3.83 s


[Document(page_content="The chatbots are in their respective folders with an application launcher `autorun.sh` to install each of them without specific knowledge.  \nOn Mac, open the terminal and type:\n```shell\ncd\n```\nDrag the **`folder`** containing the file `autorun.sh`, then press the Enter key (↩︎).  \n_If you have done it correctly, the **`~`** between your machine's name (`name@MacBook-Pro-of-Name`) and the **`%`** sign should display the name of the `folder` instead._  \nExecute the following line of code by pressing the Enter key (↩︎):\n```shell\nsh autorun.sh\n```\nWait a moment, the model should open in your default web browser.", metadata={'vector': [-0.00478768115863204, -0.04618072137236595, 0.02846730872988701, -0.03609825298190117, 0.03550594300031662, -0.03747408464550972, -0.037277232855558395, 0.04082599654793739, 0.046806637197732925, 0.020041460171341896, 0.04236359894275665, -0.03443154692649841, 0.02185412310063839, 0.03714746609330177, 0.09091991931200027, 0.

## Qdrant

In [34]:
%%time 

# Store
from langchain_community.vectorstores import Qdrant
docsearch = Qdrant.from_documents(
    docs,
    embeddings,
    path="/tmp/local_qdrant",
    collection_name="my_documents",
)

CPU times: user 22.7 s, sys: 584 ms, total: 23.3 s
Wall time: 5.87 s


In [35]:
%%time
for i in range(repeat):
    docsearch.similarity_search(query)

docsearch.similarity_search(query)

CPU times: user 19.3 s, sys: 3.62 s, total: 22.9 s
Wall time: 3.99 s


[Document(page_content="The chatbots are in their respective folders with an application launcher `autorun.sh` to install each of them without specific knowledge.  \nOn Mac, open the terminal and type:\n```shell\ncd\n```\nDrag the **`folder`** containing the file `autorun.sh`, then press the Enter key (↩︎).  \n_If you have done it correctly, the **`~`** between your machine's name (`name@MacBook-Pro-of-Name`) and the **`%`** sign should display the name of the `folder` instead._  \nExecute the following line of code by pressing the Enter key (↩︎):\n```shell\nsh autorun.sh\n```\nWait a moment, the model should open in your default web browser.", metadata={'Header 1': 'Onboarding Bot Model', 'Header 2': 'Installation', '_id': '04a4e6b37b0948829db79f436d5409fd', '_collection_name': 'my_documents'}),
 Document(page_content="The chatbots are in their respective folders with an application launcher `autorun.sh` to install each of them without specific knowledge.  \nOn Mac, open the terminal 

## SKLearn

In [36]:
%%time 

# Store
from langchain_community.vectorstores import SKLearnVectorStore
docsearch = SKLearnVectorStore.from_documents(docs, embeddings)

CPU times: user 20.7 s, sys: 441 ms, total: 21.2 s
Wall time: 4.86 s


In [37]:
%%time
for i in range(repeat):
    docsearch.similarity_search(query)

docsearch.similarity_search(query)

CPU times: user 19.4 s, sys: 3.68 s, total: 23 s
Wall time: 3.9 s


[Document(page_content="The chatbots are in their respective folders with an application launcher `autorun.sh` to install each of them without specific knowledge.  \nOn Mac, open the terminal and type:\n```shell\ncd\n```\nDrag the **`folder`** containing the file `autorun.sh`, then press the Enter key (↩︎).  \n_If you have done it correctly, the **`~`** between your machine's name (`name@MacBook-Pro-of-Name`) and the **`%`** sign should display the name of the `folder` instead._  \nExecute the following line of code by pressing the Enter key (↩︎):\n```shell\nsh autorun.sh\n```\nWait a moment, the model should open in your default web browser.", metadata={'id': 'a301206c-d425-4ef0-90ae-66415b98b4c8', 'Header 1': 'Onboarding Bot Model', 'Header 2': 'Installation', '_id': '8d2197ece10e45968e31efbf36ac3f50', '_collection_name': 'my_documents'}),
 Document(page_content='The [Dev Launchers](https://devlaunchers.org) website is a community of aspiring developers eager to learn and gain experi

## SQLite-VSS

In [38]:
%%time 

# Store
from langchain_community.vectorstores import SQLiteVSS
docsearch = SQLiteVSS.from_documents(docs, embeddings)

CPU times: user 23 s, sys: 609 ms, total: 23.7 s
Wall time: 5.72 s


In [39]:
%%time
for i in range(repeat):
    docsearch.similarity_search(query)

docsearch.similarity_search(query)

CPU times: user 23.7 s, sys: 3.72 s, total: 27.4 s
Wall time: 4.81 s


[Document(page_content="The chatbots are in their respective folders with an application launcher `autorun.sh` to install each of them without specific knowledge.  \nOn Mac, open the terminal and type:\n```shell\ncd\n```\nDrag the **`folder`** containing the file `autorun.sh`, then press the Enter key (↩︎).  \n_If you have done it correctly, the **`~`** between your machine's name (`name@MacBook-Pro-of-Name`) and the **`%`** sign should display the name of the `folder` instead._  \nExecute the following line of code by pressing the Enter key (↩︎):\n```shell\nsh autorun.sh\n```\nWait a moment, the model should open in your default web browser.", metadata={'Header 1': 'Onboarding Bot Model', 'Header 2': 'Installation', '_id': 'b7c7c2bfc87940c5af5a7d5c1b3edec8', '_collection_name': 'my_documents'}),
 Document(page_content="The chatbots are in their respective folders with an application launcher `autorun.sh` to install each of them without specific knowledge.  \nOn Mac, open the terminal 

## TileDB

In [40]:
# !brew install typesense/tap/typesense-server@0.25.2
!brew services start typesense-server@0.25.2

To reinstall 0.25.2, run:
  brew reinstall typesense-server@0.25.2
Service `typesense-server@0.25.2` already started, use `brew services restart typesense-server@0.25.2` to restart.


In [41]:
%%time 

# Store
from langchain_community.vectorstores import Typesense
docsearch = Typesense.from_documents(
    docs,
    embeddings,
    typesense_client_params={
        "host": "localhost",  # Use xxx.a1.typesense.net for Typesense Cloud
        "port": "8108",  # Use 443 for Typesense Cloud
        "protocol": "http",  # Use https for Typesense Cloud
        "typesense_api_key": "xyz",
        "typesense_collection_name": "lang-chain",
    },
)

CPU times: user 22 s, sys: 878 ms, total: 22.9 s
Wall time: 5.59 s


In [42]:
%%time
for i in range(repeat):
    docsearch.similarity_search(query)

docsearch.similarity_search(query)

CPU times: user 10.1 s, sys: 151 ms, total: 10.2 s
Wall time: 3.42 s


[Document(page_content='Repository for Mincraft build by the Dev Launchers community', metadata={'Header 1': 'community-minecraft'}),
 Document(page_content='Repository for Mincraft build by the Dev Launchers community', metadata={'Header 1': 'community-minecraft'}),
 Document(page_content='Repository for Mincraft build by the Dev Launchers community', metadata={'Header 1': 'community-minecraft'}),
 Document(page_content='Recipe book', metadata={'Header 1': 'project__recipe-book'}),
 Document(page_content='Recipe book', metadata={'Header 1': 'project__recipe-book'}),
 Document(page_content='Recipe book', metadata={'Header 1': 'project__recipe-book'}),
 Document(page_content='Merging to release branch will deploy to production server.', metadata={'Header 1': 'Deployment', 'Header 2': 'Production'}),
 Document(page_content='Merging to release branch will deploy to production server.', metadata={'Header 1': 'Deployment', 'Header 2': 'Production'}),
 Document(page_content='Merging to relea

## USearch

In [43]:
%%time 

# Store
from langchain_community.vectorstores import USearch
docsearch = USearch.from_documents(docs, embeddings)

CPU times: user 22.1 s, sys: 523 ms, total: 22.6 s
Wall time: 5.43 s


In [44]:
%%time
for i in range(repeat):
    docsearch.similarity_search(query)

docsearch.similarity_search(query)

CPU times: user 9.66 s, sys: 111 ms, total: 9.77 s
Wall time: 2.61 s


[Document(page_content="The chatbots are in their respective folders with an application launcher `autorun.sh` to install each of them without specific knowledge.  \nOn Mac, open the terminal and type:\n```shell\ncd\n```\nDrag the **`folder`** containing the file `autorun.sh`, then press the Enter key (↩︎).  \n_If you have done it correctly, the **`~`** between your machine's name (`name@MacBook-Pro-of-Name`) and the **`%`** sign should display the name of the `folder` instead._  \nExecute the following line of code by pressing the Enter key (↩︎):\n```shell\nsh autorun.sh\n```\nWait a moment, the model should open in your default web browser.", metadata={'Header 1': 'Onboarding Bot Model', 'Header 2': 'Installation', '_id': '8d2197ece10e45968e31efbf36ac3f50', '_collection_name': 'my_documents'}),
 Document(page_content='The [Dev Launchers](https://devlaunchers.org) website is a community of aspiring developers eager to learn and gain experience. As such, projects are open-source and am

# Result 

| Embeddings Module     | Speed Create table (CPU) | Speed Search (CPU) | Good Answer      |
|-----------------------|--------------------------|--------------------|:----------------:|
| BGE                   |52.3 s                    |1.01 s              |🟠                |
| FastEmbed             |1min 11s                  |1.20 s              |🟠                |
| GPT4All               |22.9 s                    |1.18 s              |🟢                |
| Hugging Face          |1min 27s                  |2.15 s              |🟢                |
| ModelScope            |37.7 s                    |1.97 s              |🟢                |
| OpenClip              |                          |                    |🔴                |
| HuggingFaceEmbeddings |1min 23s                  |2.10 s              |🟢                |
| SpaCy                 |                          |                    |🔴                |